### PR TITLE
Enable debug symbols during compilation

### PR DESCRIPTION
**Pull request by Leoquicenoz**
|Related issue|
|---|
|closes https://github.com/wazuh/wazuh/issues/21731|

## Description

Enable compilation with debug symbols while maintaining optimizations for Linux, Windows, and macOS platforms

The `CMakeLists.txt` files are modified so that the compile flag is applied by default: 

```
if(NOT CMAKE_BUILD_TYPE)
  set(CMAKE_BUILD_TYPE RelWithDebInfo)
endif()
``` 
## Tests

The compilation of the manager is carried out without errors, an error-free compilation of the Linux, Windows and Mac agents is also obtained, it is observed that the flags used in the compilation change and the resulting binaries are evaluated whit `gdb` in each case, verifying that their size increases and that debug information is obtained.

- Windows output

```
File win32/win_agent.c:
35:	void agent_help();
48:	int main(int, char **);

File win32/win_service.c:
103:	int CheckServiceRunning();
132:	int InstallService(char *);
253:	void OssecServiceCtrlHandler@4(DWORD);
307:	void OssecServiceStart@8(DWORD, LPTSTR *);
208:	int UninstallService();
285:	void WinSetError();
291:	int os_WinMain(int, char **);
39:	int os_start_service();
68:	int os_stop_service();

Non-debugging symbols:
0x004014f0  __gcc_register_frame
0x004015a0  __gcc_deregister_frame
0x004c0a5c  yaml_parser_set_input_file
0x004c0a6c  yaml_parser_load
0x004c0a7c  yaml_parser_initialize
0x004c0a8c  yaml_parser_delete
0x004c0a9c  yaml_document_get_root_node
0x004c0aac  yaml_document_get_node
0x004c0abc  yaml_document_delete
0x004c0acc  uncompress
0x004c0adc  pcre2_match_data_free_8
0x004c0aec  pcre2_match_data_create_from_pattern_8
0x004c0afc  pcre2_match_8
0x004c0b0c  pcre2_get_ovector_pointer_8
0x004c0b1c  pcre2_compile_8
0x004c0b2c  pcre2_code_free_8
0x004c0b3c  gzwrite
0x004c0b4c  gzread
0x004c0b5c  gzopen
0x004c0b6c  gzerror
``` 
image

-  MacOS output 

```
File wazuh_modules/wm_syscollector.c:
45:	const wm_context WM_SYS_CONTEXT;
36:	const char *WM_SYS_LOCATION;
41:	_Bool need_shutdown_wait;
70:	int queue_fd;
43:	_Bool shutdown_process_started;
42:	pthread_mutex_t sys_reconnect_mutex;
39:	pthread_cond_t sys_stop_condition;
40:	pthread_mutex_t sys_stop_mutex;
55:	void *syscollector_module;
56:	syscollector_start_func syscollector_start_ptr;
57:	syscollector_stop_func syscollector_stop_ptr;
69:	long syscollector_sync_max_eps;
58:	syscollector_sync_message_func syscollector_sync_message_ptr;

File wazuh_modules/wmodules.c:
19:	int wm_task_nice;
18:	wmodule *wmodules;
30:	static const void *default_modules[2];
20:	static gid_t wm_gid;

Non-debugging symbols:
0x00000001000cc000  Read_Authd.xml_disabled
0x00000001000cc008  Read_Authd.xml_port
0x00000001000cc010  Read_Authd.xml_ipv6
0x00000001000cc070  Read_Authd.xml_ssl_manager_key
0x00000001000cc0d0  Read_Cluster.key
0x00000001000cc0d8  Read_Cluster.socket_timeout
0x00000001000cc0e0  Read_Cluster.connection_timeout
0x00000001000cc0e8  Read_Cluster.interval
0x00000001000cc0f0  Read_Cluster.nodes
0x00000001000cc0f8  Read_Cluster.hidden
0x00000001000cc100  Read_Cluster.port
0x00000001000cc108  Read_Cluster.bind_addr
0x00000001000cc110  Read_Cluster.C_VALID
0x00000001000cc170  get_paths_from_env_variable[DELIM]
0x00000001000cc82c  wm_sca_read_files.first_scan
0x00000001000ccb00  OS_BF_Str.cbc_iv
0x00000001000ccb10  OS_AES_Str.iv
0x00000001000ccbc8  randombytes[fh]
0x00000001000ccc80  get_unix_version[pattern]
0x00000001000ccc88  get_unix_version.pattern

``` 
Screenshot from 2024-02-20 18-23-43

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] MAC OS X
- [x] Source installation

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -358,14 +358,8 @@ ifneq (,$(filter ${DEBUGAD},YES yes y Y 1))
...
-ifneq (,$(filter ${DEBUG},YES yes y Y 1))
-	OSSEC_CFLAGS+=-g
-	AR_LDFLAGS+=-g
-else
-	OSSEC_CFLAGS+=-DNDEBUG
-	OFLAGS+=-O2
-	AR_LDFLAGS+=-s
-endif #DEBUG
+OSSEC_CFLAGS+=-g
+AR_LDFLAGS+=-g
```
Comment (tdrauncieal) [inline src/Makefile:363]: Please, explain why all optimizations are stripped.

Comment (Leoquicenoz) [reply src/Makefile:363]: In the tests performed while modifying the code, it was found that by changing only `-DCMAKE_BUILD_TYPE=RelWithDebInfo`, but keeping the optimizations, no modification was made in the binaries and the compilation was practically the same as if you had `-DCMAKE_BUILD_TYPE=Release` (which is the type of compilation that is currently the default), tests were performed removing the optimizations one by one and see the result, but the only one where we had changes was with the configuration:

```
OSSEC_CFLAGS+=-g
AR_LDFLAGS+=-g
``` 
It was also compared with the resulting binaries when the compilation is done with `DEBUG=1` and it was evidenced that the size of the resulting binaries in this case are of a larger size (which is expected).

File: src/Makefile [hunk 2]
```diff
@@ -362,9 +362,8 @@ ifneq (,$(filter ${DEBUG},YES yes y Y 1))
 	OSSEC_CFLAGS+=-g
```
Comment (tdrauncieal) [inline src/Makefile:362]: Why are you stripping the `-DNDEBUG` flag?
also consider this code simplification of entire `if` block:
```makefile
OSSEC_CFLAGS+=-g
AR_LDFLAGS+=-g

ifeq (,$(filter ${DEBUG},YES yes y Y 1))
	OSSEC_CFLAGS+=-DNDEBUG
	OFLAGS+=-O2
	AR_LDFLAGS+=-s
endif
```

Comment (Leoquicenoz) [reply src/Makefile]: Done in [Restore optimizations](https://github.com/wazuh/wazuh/pull/22011/commits/a22e559773120d81709285d5710864269afa0923) commit

**Review by tdrauncieal — APPROVED** [review]
LGTM!

**Review by ncvicchi — APPROVED** [review]
LGTM!

**Pull request merged**